packages/tre/src/plugins/r2/gateway.ts

```diff
@@ -1,24 +1,537 @@
+  onlyIf?: R2Conditional | Headers;
+  // Various HTTP headers associated with the object. Refer to
+  // https://developers.cloudflare.com/r2/runtime-apis/#http-metadata.
+  httpMetadata?: R2HTTPMetadata | Headers;
+  // A map of custom, user-defined metadata that will be stored with the object.
+  customMetadata?: Record<string, string>;
+  // A md5 hash to use to check the recieved object’s integrity.
+  md5?: ArrayBuffer | string;
+}
+
+export type R2ListOptionsInclude = ("httpMetadata" | "customMetadata")[];
+
+export interface R2ListOptions {
+  // The number of results to return. Defaults to 1000, with a maximum of 1000.
+  limit?: number;
+  // The prefix to match keys against. Keys will only be returned if they start with given prefix.
+  prefix?: string;
+  // An opaque token that indicates where to continue listing objects from.
+  // A cursor can be retrieved from a previous list operation.
+  cursor?: string;
+  // Key after which the list results should start, exclusive.
+  startAfter?: string;
+  // The character to use when grouping keys.
+  delimiter?: string;
+  // Can include httpMetadata and/or customMetadata. If included, items returned by
+  // the list will include the specified metadata. Note that there is a limit on the
+  // total amount of data that a single list operation can return.
+  // If you request data, you may recieve fewer than limit results in your response
+  // to accomodate metadata.
+  // Use the truncated property to determine if the list request has more data to be returned.
+  include?: R2ListOptionsInclude;
+}
+
+export interface R2Objects {
+  // An array of objects matching the list request.
+  objects: R2Object[];
+  // If true, indicates there are more results to be retrieved for the current list request.
+  truncated: boolean;
+  // A token that can be passed to future list calls to resume listing from that point.
+  // Only present if truncated is true.
+  cursor?: string;
+  // If a delimiter has been specified, contains all prefixes between the specified
+  // prefix and the next occurence of the delimiter.
+  // For example, if no prefix is provided and the delimiter is ‘/’, foo/bar/baz
+  // would return foo as a delimited prefix. If foo/ was passed as a prefix
+  // with the same structure and delimiter, foo/bar would be returned as a delimited prefix.
+  delimitedPrefixes: string[];
+}
+
+const MAX_LIST_KEYS = 1_000;
+const MAX_KEY_SIZE = 1024;
+// https://developers.cloudflare.com/r2/platform/limits/ (5GB - 5MB)
+const MAX_VALUE_SIZE = 5 * 1_000 * 1_000 * 1_000 - 5 * 1_000 * 1_000;
+const UNPAIRED_SURROGATE_PAIR_REGEX =
+  /^(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])$/;
+
+const encoder = new TextEncoder();
+
+type Method = "HEAD" | "GET" | "PUT" | "LIST" | "DELETE";
+
+function throwR2Error(method: Method, status: number, message: string): void {
+  throw new Error(`R2 ${method} failed: (${status}) ${message}`);
+}
```

Comment:
Fixed (kinda) in the latest push. I have a subclass of `Error` `export class R2Error extends Error`, which encodes the status code & CF error code (I didn't actually think of subclassing HttpError, primarily because each error needs explicitly caught and converted to a response, as the runtime needs each error formatted fairly specifically)